TODO/yammer-ios-app-ported-to-swift-3.md

```diff
@@ -1,65 +1,68 @@
+*   **Enum 例子转成小写**。有些枚举变成了首字母小写，有些不会。所以，我们手动做了这部分改动。这个迁移工具会把那些有敏感词冲突的单词，比如 `default` 通过使用逆向大小写来处理。
+*   **你真的是可选的?** 有些 API 改变了，采用了可选类型。如果这个一个内部的 Object C 的 API 的话，确保你的可空标识是正确设置了。
+*   **Objective C 可空标识符**。在 Swift 3 中，每个导入的没有可空标识符的 Objective C 类都会被强制解包到可选。
+最快的解决方法是每个地方都用 `if let` 或者 `gurad let`，但是在这么做之前，在 Objective C 这边做个检查。
+*   **Optional 可比性**。因为一些 API 中的可选性的改变，并且事实上也有许多 Objective C 的 API 的改变（见上面），迁移工具会为泛型的可选类型生成一些比较函数（ `func < (lhs: T?, rhs: T?) -> Bool` ），这是一个坏的点子，很可能你的逻辑需要改变，一些代码也要删除。
+*   **NSNumber!**。Swift 3 不再会自动桥接 number 到 `NSNumber` 了（或者相似的其他 NS 类），但是在大部分的例子中，这个是不需要强制的。把它们都检查一遍。
+*   **DispatchQueue**。我喜欢这个新的 DispatchQueue 语法，但是迁移工具把一些转换搞混了。并且代码中的每一个 `dispatchAfter` 必须检查一遍方式重复转换到纳秒。因为大部分的 API 会用秒级的延迟，我们通过 `NSEC_PER_SEC` 来执行乘法加倍的操作，而迁移工具会使用这个逻辑并且通过 `NSEC_PER_SEC` 来分割，这中解决方法不够漂亮。
+*   **NSNotification.Name**.  现在 `NotificationCenter` 不再通过 `String` 而是  `NSNotification.Name`  来增加 observer 。迁移工具会把原来的量常量包装在一个 `Notification.name 中，然而我们更倾向于把 `Notification.name`  赋给一个 let 变量来隐藏常量的逻辑。
+*   **NSRange 转变 Range**.大部分的字符串 API 现在使用 `Range` 来替换 `NSRange` 。现在通过使用 literal ranges (0..<9) 更加容易操作它们。总的来说，ranges 在 Swift 中改变了很多，每个人在使用它的时候都崩溃过。重新检查一下它们，你的代码库值得这个变换！
+*   **_ 第一个参数**。Swift 3 命名规范改变来暗指着函数的第一个参数，大部分你的 api 和 api 调用都会自动改变，有些则不会。更糟糕的是，有些建议的 api 改动导致你的函数变得难以阅读。想用 `NS_SWIFT_NAME` 作为那些 Objective C 的名字是不够 swift 化的。
+*   **Objective C 类属性**。许多类的调用在 Swift 现在通过类属性的方式来实现之前的类方法(除了: _ `UIColor.red`)。在你的 objective c 中 你可以吧一个 get 方法转换为一个[静态属性]((http://useyourloaf.com/blog/objective-c-class-properties/))，他会在两个环境下生效。
+*   **Any 和 AnyObject** Objective C id 中的 id 类型现在不再由 anyobject 而是 any 来代替了。这个转换相当容易地就能解决但是也可能导致一些行的误导，[阅读](http://kuczborski.com/2014/07/29/any-vs-anyobject/)和理解它们的不同之处。
+*   **权限控制** 我们已经讨论过 `private` 和 `filePrivate`。同样值得去重新检查一下 `open`， `public` 和 `internal` 。这是另外一个要在团队内部要达成一致协议的重要事情。
 
-**Conclusions**
+**结论**
 
-The process of migrating ~ 180 Swift files took around 2 weeks and 2 people. We decided on pair migrating (_I call dibs on the name!_) because of the specific advantages in this conditions. Having 4 eyes instead of 2 becomes even more important when the focus of the project is less about code logic and more on making sure no new bugs are introduced because of typos, rename operations and reordering. A second set of hands and a laptop are very handy to check the original code when what you see in front of you does not quite make sense. Overall, it makes a task that is not that fun more enjoyable, and when everything fails at least you can switch. Thank you Mannie ([@mannietags](https://twitter.com/mannietags)) for pairing and enduring.
+迁移的过程～ 180 个 Swift 文件花了两个人的两个星期时间。我们选择结对迁移(_我这么称呼它!_)是因为这样的条件下有特别的好处。当这个项目的重点少部分在代码逻辑，而大量的在确保没有由于打字错误，重命名操作符和重排导致的新的bug时，四只眼睛比两只眼睛变得重要的多了。当你眼前的东西没有什么意义的时候，多一双手和一台笔记本来检查原始的代码会变得非常方便。总的来说，这个任务既不有趣也不轻松，不过当所有事情都失败的时候，至少你还能切换。感谢 Mannie([@mannietags](https://twitter.com/mannietags)) 的陪伴和忍耐。
```

Comment:
Having 4 eyes instead of 2 becomes even more important when the focus of the project is less about code logic and more on making sure no new bugs are introduced because of typos, rename operations and reordering.

有四只而不是两只眼睛来确保 ....  显得尤为重要。